features/include/pcl/features/impl/scurv.hpp

```diff
@@ -0,0 +1,864 @@
+#include <string.h>
+#include <math.h>
+#include <numeric>
+#include <algorithm>
+#include <functional>
+#include <typeinfo>
+#include <stdio.h>
+#include <stdlib.h>
+#include <utility>
+
+struct theMeanStruct {
+  double theMeanX;
+  double theMeanY;
+  double theMeanZ;
+} theMeanStruct;
+
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+template <typename PointInT, typename PointNT, typename PointOutT> bool
+pcl::SCurVEstimation<PointInT, PointNT, PointOutT>::compare (
+    int index1, int index2, std::vector<double> &data)
+{
+  return data[index1] < data[index2];
+}
+
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+template <typename PointInT, typename PointNT, typename PointOutT> double
+pcl::SCurVEstimation<PointInT, PointNT, PointOutT>::uvalue (double x, double low, double high)
+{
+  return (x - low) / (high - low);
+}
+
+//////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+template <typename PointInT, typename PointNT, typename PointOutT> double
+pcl::SCurVEstimation<PointInT, PointNT, PointOutT>::HermiteDerivativeInterpolate (double x1, double x2, double f1, double f2, double d1, double d2, double xi)
+{
+  double h, delta, del1, del2, c2, c3, x;
+  h = x2 - x1;
+  delta = ( f2 - f1 ) / h;
+  del1 = ( d1 - delta ) / h;
+  del2 = ( d2 - delta ) / h;
+  c2 = -( del1 + del1 + del2 );
+  c3 = ( del1 + del2 ) / h;
+  x = xi - x1;
+  return f1 + x * ( d1 + x * ( c2 + x * c3 ) );
+}
+
+//////////////////////////////////////////////////////////////////////////////////////////////
+template <typename PointInT, typename PointNT, typename PointOutT> void
+pcl::SCurVEstimation<PointInT, PointNT, PointOutT>::normalize2D (pcl::PointCloud<pcl::PointNormal> &cloudToBeNormalized, int min_range, int max_range)
+{
+  Eigen::Vector4f min_pt;
+  Eigen::Vector3f min_coeff;
+  Eigen::Vector4f max_pt;
+  Eigen::Vector3f max_coeff;
+  double minimum, maximum;
+  
+  pcl::getMinMax3D(cloudToBeNormalized, min_pt, max_pt);
+  min_coeff << min_pt(0), min_pt(1), min_pt(2);
+  max_coeff << max_pt(0), max_pt(1), max_pt(2);
+  minimum = min_coeff.minCoeff();
+  maximum = max_coeff.maxCoeff();
+  
+  double factor = fabs(minimum - maximum) / (max_range - min_range);
+//  //std::cout << "factor: " << factor << std::endl;
+  size_t k;
+  for (k = 0; k < cloudToBeNormalized.points.size(); k++)
+  {
+    cloudToBeNormalized.points[k].x = (double)cloudToBeNormalized.points[k].x / factor - maximum / factor + max_range;
+//    //std::cout << cloudToBeNormalized.points[k].x << "\t";
+    cloudToBeNormalized.points[k].y = (double)cloudToBeNormalized.points[k].y / factor - maximum / factor + max_range;
+//    //std::cout << cloudToBeNormalized.points[k].y << "\t";
+    cloudToBeNormalized.points[k].z = (double)cloudToBeNormalized.points[k].z / factor - maximum / factor + max_range;
```

Comment:
This normalization looks incorrect. The fact that your summing `max_range` at the end is suspicious.